src/http/Stream.cc

```diff
@@ -263,11 +262,15 @@ Http::Stream::socketState()
 }
 
 void
-Http::Stream::sendStartOfMessage(HttpReply *rep, StoreIOBuffer bodyData)
+Http::Stream::sendStartOfMessage(const HttpReplyPointer &rep, StoreIOBuffer bodyData)
 {
-    prepareReply(rep);
     assert(rep);
-    MemBuf *mb = rep->pack();
+    reply = rep;
```

Comment:
This HttpReply object may be passed from its raw/unprotected/lockless DeferredParams::reply storage to now-refconted Http::Stream::reply storage. I hope those deferred parameters always belong to the same Http::Stream object, but if they are not, then Http::Stream::reply object might be destroyed (via now-added Http::Stream::reply reference counting) before the corresponding DeferredParams object is destroyed, creating a dangling pointer inside DeferredParams storage. I recommend upgrading Http::Stream::DeferredParams::reply _together_ with Http::Stream::reply (i.e. in this PR) to mitigate that risk.

Reply:
Looking at branch commit 751624939712c530279837ce6dc62766e69093ed (and, to a degree, its parent), it feels like something went wrong while upgrading Http::Stream::DeferredParams::reply storage. If possible, please avoid changing all (or the vast majority of) those functions! During these transitions, it is OK to pass a raw pointer (to a reference-counted HttpReply object) to legacy code that does not store that pointer or stores it properly. AFAICT, most of those modified functions fall into the former category. They should not be modified.

Reply:
There is a chain reaction going on here. This is as small a chain of side effects as I can make it.

A) To fix the dangling pointer in `Http::Stream::DeferredParams::reply` update `Http::Stream::DeferredParams::reply` to RefCount.

That pushes the dangling pointer problem into ESIIncludeContext and Downloader where class methods using `CSCB` type for AsyncCall parameters as a raw-pointer (now the dangling one).

B) To fix (A) update `CSCB` definition to pass RefCount instead. Which pulls in all the `CSCB` functions.

Reply:
> There is a chain reaction going on here. This is as small a chain of side effects as I can make it.

Thank you for detailing this! I hope we can find a way to stop that chain reaction much sooner than branch commit [7516249](https://github.com/squid-cache/squid/pull/1855/commits/751624939712c530279837ce6dc62766e69093ed) does:
 
> A) To fix the dangling pointer in `Http::Stream::DeferredParams::reply` update `Http::Stream::DeferredParams::reply` to RefCount.
> 
> That pushes the dangling pointer problem into ESIIncludeContext and Downloader where class methods using `CSCB` type for AsyncCall parameters as a raw-pointer (now the dangling one).

CSCB callbacks are synchronous, right? When I look at Downloader changes in branch commit [7516249](https://github.com/squid-cache/squid/pull/1855/commits/751624939712c530279837ce6dc62766e69093ed), I do not see any long-term HttpReply storage. I only see caller's reply pointer being forwarded to another synchronous call. That kind of forwarding, by itself, does not require smart pointers and does not create dangling raw pointers. Let's temporary assume, to simplify arguments, that downloaderRecipient() is the only CSCB callback in existence. Why do we have to change downloaderRecipient() and CSCB type after changing Http::Stream::DeferredParams::reply?

Reply:
This should be fixed now with the latest rebase.